src/nsrt_learning/strips_learning/gen_to_spec_learner.py

```diff
@@ -144,46 +130,30 @@ def _backchain_one_pass(
                         if len(param_opt_to_nec_pnads[option.parent]) == 0:
                             param_opt_to_nec_pnads[option.parent].append(pnad)
                         break
+
                 # If we weren't able to find a substitution (i.e, the above
-                # for loop did not break), we need to try specializing each
-                # of our PNADs.
+                # for loop did not break), we need to spawn a new PNAD from
+                # the most general PNAD to cover these necessary add effects.
                 else:
-                    nec_pnad_set_changed = True
-                    for pnad in pnads_for_option:
-                        new_pnad = self._try_specializing_pnad(
-                            necessary_add_effects, pnad, segment)
-                        if new_pnad is not None:
-                            assert new_pnad.option_spec == pnad.option_spec
-                            if len(param_opt_to_nec_pnads[option.parent]) > 0:
-                                param_opt_to_nec_pnads[option.parent].remove(
-                                    pnad)
-                            del pnad
-                            break
-                    # If we were unable to specialize any of the PNADs, we need
-                    # to spawn from the most general PNAD and make a new PNAD
-                    # to cover these necessary add effects.
-                    else:
-                        new_pnad = self._try_specializing_pnad(
-                            necessary_add_effects,
-                            param_opt_to_general_pnad[option.parent], segment)
-                        assert new_pnad is not None
-
-                    pnad = new_pnad
-                    del new_pnad  # unused from here
+                    pnad = self._spawn_new_pnad(
+                        necessary_add_effects,
+                        param_opt_to_general_pnad[option.parent], segment)
                     param_opt_to_nec_pnads[option.parent].append(pnad)
-
                     # Recompute datastores for ALL PNADs associated with this
                     # option. We need to do this because the new PNAD may now
                     # be a better match for some transition that we previously
                     # matched to another PNAD.
                     self._recompute_datastores_from_segments(
                         param_opt_to_nec_pnads[option.parent])
-                    # Recompute all preconditions, now that we have recomputed
-                    # the datastores.
-                    for nec_pnad in param_opt_to_nec_pnads[option.parent]:
-                        pre = self._induce_preconditions_via_intersection(
-                            nec_pnad)
-                        nec_pnad.op = nec_pnad.op.copy_with(preconditions=pre)
+                    # If any PNAD now has an empty datastore, delete it. Then
```

Comment:
just wondering, is this change about deleting empty datastores important? or just nice to have?

Reply:
making this a todo for later